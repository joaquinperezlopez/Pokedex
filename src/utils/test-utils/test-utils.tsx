import type { PreloadedState } from '@reduxjs/toolkit';
import { setupStore, type AppStore, type RootState } from '@store/index';
import type { RenderOptions } from '@testing-library/react-native';
import { render } from '@testing-library/react-native';
import React, { PropsWithChildren } from 'react';
import { Provider } from 'react-redux';

// this is wrapper for the render function from @testing-library/react-native
// https://redux.js.org/usage/writing-tests

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>;
  store?: AppStore;
}

export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = undefined,
    // Automatically create a store instance if no store was passed in
    store = setupStore(preloadedState),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({
    children,
  }: PropsWithChildren<Record<string, unknown>>): JSX.Element {
    return <Provider store={store}>{children}</Provider>;
  }

  // Return an object with the store and all of RTL's query functions
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) };
}
