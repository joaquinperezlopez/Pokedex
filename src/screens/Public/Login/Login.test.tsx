import { describe, expect, jest, test } from '@jest/globals';
import { PublicStackNavigationProps } from '@navigation/public/public.navigator.types';
import { act, fireEvent, screen } from '@testing-library/react-native';
import { renderWithProviders } from '@utils/test-utils/test-utils';
import React from 'react';
import LoginScreen from './Login.screen';

const mockNavigationProps = {
  navigation: {
    navigate: jest.fn(),
    goBack: jest.fn(),
    reset: jest.fn(),
    setParams: jest.fn(),
    dispatch: jest.fn(),
    addListener: jest.fn(),
    removeListener: jest.fn(),
    canGoBack: jest.fn(),
    isFocused: jest.fn(),
    dangerouslyGetParent: jest.fn(),
    dangerouslyGetState: jest.fn(),
  },
  route: {
    key: 'Login',
    name: 'Login',
    params: {
      email: '',
    },
  },
} as unknown as PublicStackNavigationProps<'Login'>;

describe('LoginScreen test suite', () => {
  test('renders correctly', () => {
    // Render the LoginScreen component with the mock route
    const { toJSON } = renderWithProviders(
      <LoginScreen {...mockNavigationProps} />
    );

    // Perform any assertions you need on the rendered tree
    expect(toJSON()).toMatchSnapshot();
  });

  test('sign in button is disabled when needed', async () => {
    renderWithProviders(<LoginScreen {...mockNavigationProps} />);

    const emailInput = await screen.findByPlaceholderText(
      'screens.signUp.email'
    );
    const passwordInput = await screen.findByPlaceholderText(
      'screens.signUp.password'
    );
    const signInButton = await screen.findByTestId(
      'primaryButton' + 'screens.login.signIn'
    );

    await act(async () => {
      fireEvent.changeText(emailInput, 'notvalidemail');

      // write a valid password
      fireEvent.changeText(passwordInput, 'validPassword');
    });

    expect(signInButton.props.accessibilityState.disabled).toBeTruthy();

    await act(async () => {
      fireEvent(emailInput, 'changeText', 'test@test.com');
    });

    expect(signInButton.props.accessibilityState.disabled).toBeFalsy();
  });
});
