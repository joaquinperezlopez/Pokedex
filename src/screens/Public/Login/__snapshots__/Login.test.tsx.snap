// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginScreen test suite renders correctly 1`] = `
<RNCSafeAreaView
  edges={
    {
      "bottom": "additive",
      "left": "additive",
      "right": "additive",
      "top": "additive",
    }
  }
  style={
    {
      "flex": 1,
      "padding": 10,
    }
  }
>
  <View
    style={
      {
        "alignItems": "center",
        "flex": 2,
        "justifyContent": "center",
      }
    }
  >
    <Text
      numberOfLines={1}
      style={
        {
          "color": "#fff",
          "fontSize": 32,
          "fontWeight": "bold",
        }
      }
    >
      screens.login.title
    </Text>
  </View>
  <RCTScrollView>
    <View>
      <View
        style={
          {
            "gap": 8,
            "height": 80,
          }
        }
      >
        <TextInput
          name="email"
          onBlur={[Function]}
          onChange={[Function]}
          onChangeText={[Function]}
          placeholder="screens.signUp.email"
          placeholderTextColor="#666666"
          style={
            {
              "backgroundColor": "#fff",
              "borderRadius": 6,
              "color": "#000",
              "fontSize": 14,
              "fontWeight": "bold",
              "marginHorizontal": 8,
              "padding": 8,
            }
          }
          value=""
        />
        <Text
          style={
            {
              "color": "#fff",
              "fontSize": 12,
              "fontWeight": "bold",
              "left": 8,
            }
          }
        />
      </View>
      <View
        style={
          {
            "gap": 8,
            "height": 80,
          }
        }
      >
        <TextInput
          name="password"
          onBlur={[Function]}
          onChange={[Function]}
          onChangeText={[Function]}
          placeholder="screens.signUp.password"
          placeholderTextColor="#666666"
          secureTextEntry={true}
          style={
            {
              "backgroundColor": "#fff",
              "borderRadius": 6,
              "color": "#000",
              "fontSize": 14,
              "fontWeight": "bold",
              "marginHorizontal": 8,
              "padding": 8,
            }
          }
          value=""
        />
        <Text
          style={
            {
              "color": "#fff",
              "fontSize": 12,
              "fontWeight": "bold",
              "left": 8,
            }
          }
        />
      </View>
      <View
        style={
          {
            "alignItems": "center",
            "flexDirection": "row",
            "gap": 12,
            "marginHorizontal": 8,
          }
        }
      >
        <RCTSwitch
          accessibilityRole="switch"
          onChange={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            [
              {
                "height": 31,
                "width": 51,
              },
              {
                "backgroundColor": "#DC0A2D",
                "borderRadius": 16,
              },
            ]
          }
          thumbTintColor="#fff"
          value={true}
        />
        <Text
          style={
            {
              "color": "#fff",
              "fontSize": 12,
              "fontWeight": "bold",
            }
          }
        >
          screens.login.rememberMe
        </Text>
      </View>
    </View>
  </RCTScrollView>
  <View
    style={
      {
        "gap": 16,
        "padding": 10,
      }
    }
  >
    <View
      accessibilityState={
        {
          "disabled": true,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        [
          {
            "alignItems": "center",
            "backgroundColor": "#DC0A2D",
            "borderColor": "#E0E0E0",
            "borderRadius": 6,
            "borderWidth": 1,
            "justifyContent": "center",
            "padding": 10,
          },
          {
            "opacity": 0.7,
          },
        ]
      }
      testID="primaryButtonscreens.login.signIn"
    >
      <Text
        style={
          {
            "color": "#fff",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        screens.login.signIn
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "justifyContent": "center",
          "opacity": 1,
        }
      }
    >
      <Text
        style={
          {
            "color": "#fff",
            "fontSize": 16,
            "fontWeight": "bold",
          }
        }
      >
        screens.login.noAccount
      </Text>
    </View>
  </View>
</RNCSafeAreaView>
`;
